### ISSUE TITLE
Use User#thredded_display_name instead of User#to_s

### ISSUE THREAD

**Issue by glebm** [thredded/thredded#398]
From https://github.com/thredded/thredded/issues/397#issuecomment-246917371

**Comment by timdiggins** [comment]
I'm on this.

**Comment by timdiggins** [comment]
Configurable `thredded_display_name` for User, with a default implementation (proc) of `send(Thredded.user_name_column) || ''`

**Comment by jayroh** [comment]
Would it be prudent to raise a custom Thredded exception if the method is not defined instead of returning an empty string? I have a feeling it could be a bit mysterious if someone sees `""` show up instead of a user's name, email, nickname, etc., and not know why or where this should be set and doesn't think to look through the config or the code.

**Comment by timdiggins** [comment]
@jayroh yes good shout -- but doesn't need to be a custom exception, could just be a custom message like `raise "User.#{Thredded.user_name_column} must not be nil: please set make sure non nil or configure Thredded.user_display_name_method"`

**Comment by glebm** [comment]
@timdiggins Sounds good, so something like this?

```ruby
def thredded_display_name
  send(Thredded.user_display_name_method).presence or fail "..."
end
```

```ruby
module Thredded
  mattr_accessor :user_display_name_method
  
  def self.user_display_name_method
    super || user_name_column
  end

  ...
```

**Comment by timdiggins** [comment]
@glebm @jayroh I started coding this already, thinking it was going to be hard work because we'd have to find all instances of to_s (including implicit uses), so best case was to change `User#to_s` in the dummy app to raise an exception to drive the change (and prevent future ones).

But then while doing it I realized that (AFAICT, still testing it) we can just provide a default `to_s` in `UserExtender` and this will be used if no `to_s` is defined on `User`, but if one is defined, it will be used.

If that is 100% the case (see below ) then we have two options:

1) just implement a default `UserExtender#to_s` and document the fact that thredded uses this behaviour -- this means we can continue to use the idiomatic and very rails-conventional uses of (implicit) `User#to_s`  in the codebase

2) go ahead and replace uses of "User#to_s" with "thredded_display_name" (using the raise trick above)

There is one case where I think adding `UserExtender#to_s` would actually override main_app implementation: where User includes a *different* concern or extends a parent base/abstract class and either one includes a to_s they want to use. Seems possible, if rare.

So the main question is do we prefer more rails idiomaticness in the codebase (option 1), or less (even unlikely) conflict with other engines/mainapp (option 2). And thinking this through (by typing) -- I think the answer is, the engine should prioritize ease of use / least surprise for the main_app developer over ease of use/least surprise for  the engine contributer.

This leads me to option 2. Hope you agree. Will be submitting a first WIP pr relatively soon
